ast/treemap/treemap.h

```diff
@@ -719,6 +719,481 @@ class TreeMap {
+            } else if (isa_tree<Hash>(what.get())) {
+                return mapHash(std::unique_ptr<Hash>(static_cast<Hash *>(what.release())), ctx);
+            } else if (isa_tree<ClassDef>(what.get())) {
+                return mapClassDef(std::unique_ptr<ClassDef>(static_cast<ClassDef *>(what.release())), ctx);
+            } else if (isa_tree<If>(what.get())) {
+                return mapIf(std::unique_ptr<If>(static_cast<If *>(what.release())), ctx);
+            } else if (isa_tree<While>(what.get())) {
+                return mapWhile(std::unique_ptr<While>(static_cast<While *>(what.release())), ctx);
+            } else if (isa_tree<Break>(what.get())) {
+                return mapBreak(std::unique_ptr<Break>(static_cast<Break *>(what.release())), ctx);
+            } else if (isa_tree<Retry>(what.get())) {
+                return mapRetry(std::unique_ptr<Retry>(static_cast<Retry *>(what.release())), ctx);
+            } else if (isa_tree<Next>(what.get())) {
+                return mapNext(std::unique_ptr<Next>(static_cast<Next *>(what.release())), ctx);
+            } else if (isa_tree<Return>(what.get())) {
+                return mapReturn(std::unique_ptr<Return>(static_cast<Return *>(what.release())), ctx);
+            } else if (isa_tree<Rescue>(what.get())) {
+                return mapRescue(std::unique_ptr<Rescue>(static_cast<Rescue *>(what.release())), ctx);
+            } else if (isa_tree<Field>(what.get())) {
+                return mapField(std::unique_ptr<Field>(static_cast<Field *>(what.release())), ctx);
+            } else if (isa_tree<Assign>(what.get())) {
+                return mapAssign(std::unique_ptr<Assign>(static_cast<Assign *>(what.release())), ctx);
+            } else if (isa_tree<Array>(what.get())) {
+                return mapArray(std::unique_ptr<Array>(static_cast<Array *>(what.release())), ctx);
+            } else if (isa_tree<Cast>(what.get())) {
+                return mapCast(std::unique_ptr<Cast>(static_cast<Cast *>(what.release())), ctx);
+            } else {
+                Exception::raise("should never happen. Forgot to add new tree kind? {}", what->nodeName());
+            }
+        } catch (SorbetException &e) {
+            Exception::failInFuzzer();
+
+            throw ReportedRubyException{e, loc};
+        }
+    }
+};
+
+class ShallowMap {
+public:
+    template <typename CTX, typename FUNC>
+    static unique_ptr<Expression> apply(CTX ctx, FUNC &func, unique_ptr<Expression> to) {
+        ShallowMapper<FUNC, CTX> walker(func);
+        try {
+            return walker.mapIt(move(to), ctx);
+        } catch (ReportedRubyException &exception) {
+            Exception::failInFuzzer();
+            if (auto e = ctx.state.beginError(exception.onLoc, core::errors::Internal::InternalError)) {
+                e.setHeader("Failed to process tree (backtrace is above)");
+            }
+            throw exception.reported;
+        }
+    }
+};
+
+} // namespace sorbet::ast
```

Comment:
Is this just the treemap implementation with some parts trimmed out?

Reply:
Yes; I've marked out the only place where they are different.